src/add.cpp

```diff
@@ -330,8 +330,9 @@ RCP<const Basic> sub(const RCP<const Basic> &a, const RCP<const Basic> &b)
     return add(a, mul(minus_one, b));
 }
 
-RCP<const Basic> add_expand(const RCP<const Add> &self)
+RCP<const Basic> add_expand(const RCP<const Basic> &self_)
 {
+    const RCP<const Add> self = rcp_static_cast<const Add>(self_);
```

Comment:
This is an alternative, but I don't like the declaration of new functions. (eg: `rcp_static_cast_Add`)
```C++
RCP<const Mul> rcp_static_cast_Mul(const RCP<const Basic> &self) {
    return rcp_static_cast<const Mul>(self);
}

RCP<const Pow> rcp_static_cast_Pow(const RCP<const Basic> &self) {
    return rcp_static_cast<const Pow>(self);
}

RCP<const Add> rcp_static_cast_Add(const RCP<const Basic> &self) {
    return rcp_static_cast<const Add>(self);
}

RCP<const Basic> expand(const RCP<const Basic> &self)
{
    typedef std::function<RCP<const Basic>(const RCP<const Basic>)> fn;
    using std::placeholders::_1;
    std::vector<fn> table;
    table.assign(100, NULL);
    table[ADD] = std::bind(add_expand, std::bind(rcp_static_cast_Add, _1));
    table[MUL] = std::bind(mul_expand, std::bind(rcp_static_cast_Mul, _1));
    table[POW] = std::bind(pow_expand, std::bind(rcp_static_cast_Pow, _1));
    fn f = table[self->get_type_code()];
    if (f == NULL) {
        return self;
    } else {
        return f(self);
    }
    return self;
}
```